File: packages/chat-typescript/src/thread.ts
import type { Identity, IdentityWire } from './identity'
import type { TenantScope } from './tenant'

export type Thread = {
  id: string
  tenant: TenantScope
  metadata: Record<string, unknown>
  createdAt: string
  updatedAt: string
}

export type ThreadWire = {
  id: string
  tenant: TenantScope
  metadata: Record<string, unknown>
  created_at: string
  updated_at: string
}

export type ThreadMember = {
  threadId: string
  identityId: string
  identity: Identity
  role: string
  addedAt: string
  addedBy: Identity
}

export type ThreadMemberWire = {
  thread_id: string
  identity_id: string
  identity: IdentityWire
  role: string
  added_at: string
  added_by: IdentityWire
}

export type ThreadPatch = {
  tenant?: TenantScope
  metadata?: Record<string, unknown>
}

File: packages/chat-python/src/rfnry_chat_protocol/thread.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from rfnry_chat_protocol.identity import Identity
from rfnry_chat_protocol.tenant import TenantScope


class Thread(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    id: str
    tenant: TenantScope = Field(default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class ThreadPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    tenant: TenantScope | None = None
    metadata: dict[str, Any] | None = None


class ThreadMember(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    thread_id: str
    identity_id: str
    identity: Identity
    role: str = "member"
    added_at: datetime
    added_by: Identity


class ThreadInvitedFrame(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    thread: Thread
    added_member: Identity
    inviter: Identity
